index.js

```diff
@@ -0,0 +1,330 @@
+  //.
+  //. Constant's sole data constructor. The [type representative][] makes
+  //. `Constant (M)` an [Applicative][]-compatible type representative if
+  //. `M` represents some monoidal type.
+  //.
+  //. ```javascript
+  //. > Constant (String) ('abc')
+  //. Constant (String) ('abc')
+  //.
+  //. > Constant (Number) (123)
+  //. Constant (Number) (123)
+  //. ```
+  function Constant(A) {
+    var prototype = {
+      /* eslint-disable key-spacing */
+      'constructor':            Constant$bound,
+      '@@show':                 Constant$prototype$show,
+      'fantasy-land/map':       Constant$prototype$map,
+      'fantasy-land/bimap':     Constant$prototype$bimap,
+      'fantasy-land/reduce':    Constant$prototype$reduce,
+      'fantasy-land/traverse':  Constant$prototype$traverse
+      /* eslint-enable key-spacing */
+    };
+
+    var custom = util.inspect.custom;
+    /* istanbul ignore else */
+    if (typeof custom === 'symbol') {
+      prototype[custom] = Constant$prototype$show;
+    } else {
+      prototype.inspect = Constant$prototype$show;
+    }
+
+    function Constant$bound(value) {
+      var identity = Object.create (prototype);
+      if (Z.Setoid.test (value)) {
+        identity['fantasy-land/equals'] = Constant$prototype$equals;
+        if (Z.Ord.test (value)) {
+          identity['fantasy-land/lte'] = Constant$prototype$lte;
+        }
+      }
+      if (Z.Semigroup.test (value)) {
+        identity['fantasy-land/concat'] = Constant$prototype$concat;
+        identity['fantasy-land/ap'] = Constant$prototype$ap;
+      }
+      identity.value = value;
+      return identity;
+    }
+
+    Constant$bound.toString = function() {
+      if (A['@@type'] === Constant$bound['@@type']) {
+        return 'Constant (' + show (A) + ')';
+      } else if (Object.prototype.hasOwnProperty.call (A, 'name')) {
+        return 'Constant (' + A.name + ')';
+      } else {
+        var source = String (A);
+        var match = /^\s*function ([$_A-Za-z][$_A-Za-z0-9]*)/.exec (source);
+        return 'Constant (' + (match == null ? source : match[1]) + ')';
+      }
+    };
+
+    //# Constant.@@type :: String
+    //.
+    //. Constant [type identifier][].
+    //.
+    //. ```javascript
+    //. > type (Constant (Number) (42))
+    //. 'sanctuary-constant/Constant@1'
+    //.
+    //. > type.parse (type (Constant (Number) (42)))
+    //. {namespace: 'sanctuary-constant', name: 'Constant', version: 1}
+    //. ```
+    Constant$bound['@@type'] = 'sanctuary-constant/Constant@1';
+
+    //# Constant.fantasy-land/of :: Monoid m => a -> Constant m a
+    //.
+    //. `of (Constant (M)) (x)` is equivalent to `Constant (M) (empty (M))`.
+    //.
+    //. ```javascript
+    //. > S.of (Constant (Array)) (42)
+    //. Constant (Array) ([])
+    //.
+    //. > S.of (Constant (String)) (42)
+    //. Constant (String) ('')
+    //. ```
+    (function() {
+      var empty;
+      try { empty = Z.empty (A); } catch (err) { return; }
```

Comment:
What kinds of errors can happen here? Could the attachment of `['fantasy-land/of']` be done in the `Constant$bound` function with something like `Z.Monoid.test`?

Reply:
Ah, I guess it can't be done in `Constant$bound` itself, since you need it before you have a value, but couldn't it nevertheless use `Z.Monoid.test(A)`?

Reply:
`Z.Monoid.test` takes a value rather than a type representative:

```javascript
> Z.Monoid.test ('abc')
true

> Z.Monoid.test (String)
false
```

> What kinds of errors can happen here?

```javascript
> Z.empty (Number)
TypeError: Monoid.methods.empty(...) is not a function
```

I would like to avoid `try...catch`, but I can't see how to do so.

Reply:
I guess it's off topic here, but do you think it would be a good idea to add an equivalent that did work on the type rep? Then if you have some function that can take a value and obtain it's type rep, and a function that can take a type rep and determine whether it implements a typeclass, their composition is `test`.

Reply:
That would for example allow you to avoid using the `try`/`catch` here; all the class membership testing can be done on the `A` typerep outside of the `Constant$bound` constructor, and then in `Constant$bound` you just need to test whether the typerep of `value` is indeed `A`.

Reply:
Great idea! I'll investigate changes to sanctuary-type-classes which would enable this.

Reply:
Not 100% sure, but this approach might also yield some minor performance improvements elsewhere, by moving some checking out of the path where values are being moved around.

Reply:
Has an issue been posted in sanctuary-type-classes to keep track of this?

Reply:
> Then if you have some function that can take a value and obtain [its] type rep, and __a function that can take a type rep and determine whether it implements a typeclass__, their composition is `test`.

fantasyland/fantasy-land#315 is the ultimate solution to this problem, I believe. Implementing that proposal is near the top of my list of Sanctuary-related tasks. :)